### PR TITLE
feat: DAO Escrow contract

### PR DESCRIPTION
**Pull request by netpoe**

### CODE REVIEW

File: rust-escrow/dao-escrow/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        let context = get_context(false);
+        testing_env!(context);
+
+        let contract = DaoEscrow::new("sputnikv2.testnet".parse::<AccountId>().unwrap());
+        
+        assert_eq!(
+            contract.get_dao_config("daoname".to_string(), "poguz.testnet".to_string()),
+            format!(r#"{{ "policy": {{ "roles": [ {{ "name": "Everyone", "kind": {{ "Group": [ {} ] }}, "permissions": [ "*:Finalize", "*:AddProposal", "*:VoteApprove", "*:VoteReject", "*:VoteRemove" ], "vote_policy": {{}} }}, {{ "name": "all", "kind": "Everyone", "permissions": [ "*:AddProposal" ], "vote_policy": {{}} }} ], "default_vote_policy": {{ "weight_kind": "RoleWeight", "quorum": "0", "threshold": [ 1, 2 ] }}, "proposal_bond": "100000000000000000000000", "proposal_period": "604800000000000", "bounty_bond": "100000000000000000000000", "bounty_forgiveness_period": "604800000000000" }}, "config": {{ "name": "{}", "purpose": "", "metadata": "" }} }}"#, "poguz.testnet".to_string(), "daoname".to_string()).into_bytes()
+        );
+    }
```
Comment (netpoe) [inline rust-escrow/dao-escrow/src/lib.rs:138]: I think the content of this test goes below, in `test_get_dao_config`

File: rust-escrow/dao-escrow/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+    fn get_context(is_view: bool) -> VMContext {
+        VMContextBuilder::new()
+            .signer_account_id(bob())
+            .is_view(is_view)
+            .build()
+    }
+
+    #[test]
+    fn test_daos_of() {
```
Comment (netpoe) [inline rust-escrow/dao-escrow/src/lib.rs:128]:
```suggestion
    fn test_get_daos_by_country_count() {
```

File: rust-escrow/conditional-escrow/src/lib.rs [hunk 3]
```diff
@@ -127,7 +142,15 @@ impl ConditionalEscrow {
         let payee = self.get_recipient_account_id();
         let total_funds = self.get_total_funds();
 
-        Promise::new(payee.clone()).transfer(total_funds);
+        // Create Dao
+        ext_dao_escrow::create_dao(
+            "mx".to_string(),           // country_code
+            self.deposits.to_vec(),     // deposits
+            payee.clone(),              // Contract ID
+            self.get_total_funds(),     // funds
```
Comment (netpoe) [inline rust-escrow/conditional-escrow/src/lib.rs:150]: `total_funds` is already declared above.

Comment (netpoe) [reply rust-escrow/conditional-escrow/src/lib.rs:150]: I'm assuming that passing `total_funds` becomes the `env::attached_deposit`?

File: rust-escrow/conditional-escrow/src/lib.rs [hunk 4]
```diff
@@ -127,7 +142,15 @@ impl ConditionalEscrow {
         let payee = self.get_recipient_account_id();
```
Comment (netpoe) [inline rust-escrow/conditional-escrow/src/lib.rs:142]: Let's change this to: 

```suggestion
        let recipient_contract_id = self.get_recipient_account_id();
```

File: rust-escrow/dao-escrow/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+// define the methods we'll use on the other contract
+#[ext_contract(ext_dao_factory)]
+pub trait DaoFactory {
+    fn create(&mut self, name: String, args: Base64VecU8);
+}
+
+// define methods we'll use as callbacks on our contract
+#[ext_contract(ext_self)]
+pub trait MyContract {
```
Comment (netpoe) [inline rust-escrow/dao-escrow/src/lib.rs:19]: `MyContract`?

Shouldn't it be `DaoEscrow`?

File: rust-escrow/dao-escrow/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+impl Default for DaoEscrow {
+    fn default() -> Self {
+        env::panic_str("DaoEscrow should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl DaoEscrow {
```
Comment (netpoe) [inline rust-escrow/dao-escrow/src/lib.rs:37]: I'm a bit hesitant about the name. `DaoEscrow` reads as if this works as an `Escrow` for a DAO, but it is actually a `DaoFactory`, or to be more specific `SputnikDao2Factory`. 

The difference is that this contract passes down the `deposit_accounts ` to create its members. 

Lets call the external contract trait: `ExtSputnikDao2Factory` and this one: `SputnikDao2Factory`. WDYT?

File: rust-escrow/dao-escrow/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+            GAS_FOR_CREATE_CALLBACK,
+        ))
+    }
+
+    pub fn on_create_callback(&mut self, country_code: String, daos_by_country: u128) -> bool {
+        assert_eq!(
+            env::promise_results_count(),
+            1,
+            "ERR_CALLBACK_METHOD"
+        );
```
Comment (netpoe) [inline rust-escrow/dao-escrow/src/lib.rs:100]: What is this actually asserting? just curious 🤔

**Review by netpoe — COMMENTED** [review]
@susguzman made some comments. 

Are we testing for `panic!` cases?

File: rust-escrow/dao-escrow/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+// define methods we'll use as callbacks on our contract
+#[ext_contract(ext_self)]
+pub trait MyContract {
+    fn on_create_callback(&mut self, country_code: String, daos_by_country: u128) -> bool;
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct DaoEscrow {
+    daos: UnorderedMap<String, u128>, // Country code and sequential number
```
[the remark this answers is not in the document]
Comment (susguzman) [reply rust-escrow/dao-escrow/src/lib.rs:26]: I like `daos_by_country_count`.

File: rust-escrow/dao-escrow/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        testing_env!(context);
+
+        let mut contract = DaoEscrow::new("sputnikv2.testnet".parse::<AccountId>().unwrap());
+        contract.create_dao("gt".to_string(), vec![]);
+        contract.create_dao("gt".to_string(), vec![(bob(), 1000)]);
+
+        //assert_eq!(
+        //    contract.daos_of(&"gt".to_string()),
+        //    2
+        //);
```
[the remark this answers is not in the document]
Comment (susguzman) [reply rust-escrow/dao-escrow/src/lib.rs:183]:
```suggestion
```

File: rust-escrow/conditional-escrow/src/lib.rs (hunk 3 above)
Comment (susguzman) [reply rust-escrow/conditional-escrow/src/lib.rs:150]: That's correct the `total_founds` becomes the `env::attached_deposit` for the call.

File: rust-escrow/dao-escrow/src/lib.rs (hunk 5 above)
Comment (susguzman) [reply rust-escrow/dao-escrow/src/lib.rs:19]: I had the same question but I think that my contract makes reference to this contract, the contract that your are building.

File: rust-escrow/dao-escrow/src/lib.rs (hunk 7 above)
Comment (susguzman) [reply rust-escrow/dao-escrow/src/lib.rs:100]: It's asserting that the call was made as a callback.

File: rust-escrow/dao-escrow/src/lib.rs (hunk 6 above)
Comment (susguzman) [reply rust-escrow/dao-escrow/src/lib.rs:37]: Or Maybe only `daofactory`.

Comment (netpoe) [reply rust-escrow/dao-escrow/src/lib.rs:37]: Yeah, in this case: `ExtDaoFactory` and `DaoFactory`

**Review by susguzman — APPROVED** [review]

**Pull request merged**